File: Top200/143-ReorderList.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func reorderList(head *ListNode) {
	dummy := &ListNode{Val: -1}
	dummy.Next = head
	slow := dummy
	fast := dummy

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	afterHead := reverseList(slow.Next)
	slow.Next = nil

	curr := head
	for afterHead != nil {
		temp := afterHead.Next
		afterHead.Next = curr.Next
		curr.Next = afterHead
		afterHead = temp
		curr = curr.Next.Next
	}
}

func reverseList(head *ListNode) *ListNode {
	var prev *ListNode = nil
	curr := head

	for curr != nil {
		temp := curr.Next
		curr.Next = prev
		prev = curr
		curr = temp
	}

	return prev
}

File: Top200/143-ReorderList.java
package Top200;

class ReorderList {
    class ListNode {
        int val;
        ListNode next;
        
        ListNode() {}
        
        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public void reorderList(ListNode head) {
        ListNode dummy = new ListNode(-1, head);
        ListNode slow = dummy;
        ListNode fast = dummy;
        ListNode curr = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        ListNode reverseBegin = slow.next;
        slow.next = null;
        
        reverseBegin = this.reverseList(reverseBegin);

        while (reverseBegin != null) {
            ListNode temp = reverseBegin.next;
            reverseBegin.next = curr.next;
            curr.next = reverseBegin;
            curr = reverseBegin.next;
            reverseBegin = temp;
        }
    }

    // 翻转链表
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        } 

        ListNode prev = null;
        ListNode curr = head;

        while (curr != null) {
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }

        return prev;
    }
}
